Prepare build for ES modules

jaylinski opened the pull request:
Replaced the **Browserify** build with a ES compatible **rollup.js** + **Babel 7** build.

This change is the first step in order to move from CommonJS modules to ES modules.

Related to #1031 and #1130.

## Done

* [x] If you've added code that should be tested, please add tests.
* [x] If you've modified the API (e.g. added a new config or public method), update the [docs](https://github.com/getsentry/raven-js/tree/master/docs) and TypeScript [declaration file](/getsentry/raven-js/blob/master/typescript/raven.d.ts).
* [x] Ensure your code lints
* [x] Ensure the test suite passes (npm test).

jaylinski commented on test/raven.test.js:
```diff
@@ -6,20 +6,13 @@ var sinon = require('sinon');
-  // Ensure same TraceKit obj is shared (without specifying this, proxyquire
-  // seems to clone dependencies or something weird)
-  '../vendor/TraceKit/tracekit': TraceKit
-});
```
This was probably here for a good reason.

I removed it because it depends on browserify. Also most tests seem to work without it.

kamilogorek replied:
I'm fine with removing it if all tests are passing on HeadlessChrome/HeadlessFirefox

jaylinski commented on Gruntfile.js:
```diff
@@ -85,72 +23,58 @@ module.exports = function(grunt) {
   });
 
-  var pluginCombinations = combine(plugins);
-  var pluginConcatFiles = pluginCombinations.reduce(function(dict, comb) {
```
There was probably also a reason for building all plugins in all possible combinations.

But I couldn't think of a reason why to do this, so I removed it to make the build simpler.

kamilogorek replied:
We need this functionality, as we generate all variants of a the plugin+core bundles for our hosted CDN – http://ravenjs.com/

jaylinski commented on .babelrc:
```diff
@@ -0,0 +1,7 @@
+  "presets": [
+    ["@babel/preset-env", {
+      "modules": false
+    }]
```
You probably want to define the exact targets here. See https://babeljs.io/docs/plugins/preset-env/.

kamilogorek replied:
We support all browsers from this matrix https://github.com/getsentry/raven-js
Could you please update the config reflect it?

jaylinski commented:
ICYW, the bundles got smaller. :chart_with_downwards_trend: 
``` 
browserify:       ./dist/raven.min.js: 9.80KB < maxSize 10KB gzip
rollup + babel 7: ./dist/raven.min.js: 9.96KB < maxSize 10KB gzip 
```

jaylinski commented on test/raven.test.js:
```diff
@@ -2342,13 +2335,13 @@ describe('Raven (public API)', function() {
       });
       assert.throw(function() {
-        fn.__raven__;
+        return fn.__raven__;
```
I had to add `return`, because otherwise Babel would remove this line (unassigned expression statement).

kamilogorek commented on .editorconfig:
```diff
@@ -0,0 +1,13 @@
+# http://editorconfig.org
```
We use Prettier for styling, so editorconfig is superfluous here

kamilogorek commented on Gruntfile.js:
```diff
@@ -85,72 +23,58 @@ module.exports = function(grunt) {
+      options: [
+        {
+          input: {
+            input: 'test/raven.test.js'
```
It misses all plugins and integration tests

jaylinski replied:
all files are now generated.

kamilogorek commented on Gruntfile.js:
```diff
@@ -315,34 +257,27 @@ module.exports = function(grunt) {
   ]);
-  grunt.registerTask('build', ['build.plugins-combined']);
-  grunt.registerTask('dist', ['build.core', 'copy:dist']);
+  grunt.registerTask('build', ['build.core', 'build.plugins', 'uglify']);
```
Can we utilize https://github.com/TrySound/rollup-plugin-uglify for this?

jaylinski replied:
I didn't use it because I couldn't get the plugin to output the normal **and** the uglified build.

I don't if that is possible with this plugin.

kamilogorek commented:
Thanks! I'll have to test it locally soon, so I just skimmed through the code.

btw. How's 9.96KB smaller than 9.80KB? :P

jaylinski commented:
@kamilogorek thanks for the review! I'll try to adress the issues on friday.

I mixed up the sizes (I corrected it in the comment above).

jaylinski commented:
@kamilogorek I pushed an updated commit.

* Plugin combinations are now generated again.
* I updated the `.babelrc` config to reflect your browser support matrix (although I would recommend to support the latest 2 browser versions, instead of only the latest)
* I removed the `.editorconfig` file
* All tests are now built (although the workflow is not optimal, since every new test-file has to be referenced in the `karma.unit.config.js` file. rollup is not suited for bundling all tests into a single file.) 

I also just noticed this commit: https://github.com/getsentry/raven-js/commit/5d2685fc644f62a33efd533f65f231e3a8f4cc1b
Does this commit have any implications on this PR?

kamilogorek commented:
Thanks, @jaylinski! I won't merge it before the end of the year, as most of the company (including me) is taking a break for holidays, and it's very major change. I'll test it throughout on my local env and let you know if everything is fine. If yes, then I'll merge it as a first thing in 2018  :)

> I also just noticed this commit: 5d2685f
> Does this commit have any implications on this PR?

Shhh...! it's a secret ;) No, it doesn't change anything for this PR :)

jaylinski commented:
@kamilogorek If it helps you to review, I could explain the changes I made line-by-line.

kamilogorek commented:
@jaylinski thank you for being so patient. I had a lot of things to handle recently + whole Christmas/new year time.

We tested everything separately and it's all working smooth! Thanks for your contribution, it'll become very useful soon :)

jaylinski commented:
@kamilogorek Thanks! Will you do the ES6 migration yourself? Or can/should I help?

kamilogorek commented:
Yes, we'll move to ES6/TypeScript soon, as we're working on Raven v4 :)

tommedema commented:
How's the move to es6 modules going?

kamilogorek commented:
@tommedema all our packages are available as ESM modules for quite come time now
